src/video.js

```diff
@@ -64,3 +66,47 @@ export const checkVideoBidSetup = hook('sync', function(bid, adUnit, videoMediaT
 
   return true;
 }, 'checkVideoBidSetup');
+
+export function registerVASTTrackers(tracker) {
+  vastTrackers.push(tracker);
+}
+
+export function insertVastTrackers(trackers, vastXml) {
+  const doc = new DOMParser().parseFromString(vastXml, 'text/xml');
+  const wrappers = doc.querySelectorAll('VAST Ad Wrapper, VAST Ad InLine');
+  try {
+    if (wrappers.length) {
+      wrappers.forEach(wrapper => {
+        trackers['impressions'].forEach(trackingUrl => {
+          const impression = doc.createElement('Impression');
+          impression.appendChild(doc.createCDATASection(trackingUrl));
+          wrapper.appendChild(impression)
+        });
+      });
+      vastXml = new XMLSerializer().serializeToString(doc);
+    }
+  } catch (error) {
+    logError('an error happened trying to insert trackers in vastXml');
+  }
+  return vastXml;
+}
+
+export function getVastTrackers(bid) {
+  let hasTrackers = false;
```

Comment:
this implementation needs to be cleaned, it's not easy to follow and is difficult to manage. A few things that could help:
- split into functions that do one thing and can be reused - (Single Responsibility Principle)
- work with simple data structures: it's cumbersome to work with `trackers` which is an object where each value is an array; instead you should work with arrays, and then set those arrays to a `trackers` map at the end. 
- remove flags - flags are a nuisance to maintain as the code evolves, plus they are unnecessary. If at the end of the function, the impression array is empty, you can return null. 

Lastly, the interface is hard to work with. I have to look at the implementation to understand what it's returning.
Instead, it would be easier if it returned a simpler data structure, such as a nullable object.

Reply:
Latest commit made changes which should match those you asked for. Can you tell me if there is something else you would like to see modify here or if I didn't understood what you requests ? I was not quite sure about the `trackers` map you mentioned. Thanks